File: server/internal/platform/printing/catalog.go
package printing

import (
	"fmt"
	"sort"
	"strings"

	"github.com/z1coyan/synie/server/internal/platform/apierror"
	"github.com/z1coyan/synie/server/internal/platform/meta"
)

type Field struct {
	Name  string `json:"name"`
	Label string `json:"label"`
}

type Loop struct {
	Name        string   `json:"name"`
	Label       string   `json:"label"`
	Fields      []Field  `json:"fields"`
	NestedLoops []string `json:"nestedLoops,omitempty"`
}

type ResourceCatalog struct {
	Resource string  `json:"resource"`
	Fields   []Field `json:"fields"`
	Loops    []Loop  `json:"loops"`
}

type PlaceholderSet struct {
	Fields []string            `json:"fields"`
	Nested map[string][]string `json:"nested"`
}

// FieldCatalog is the stable printing-facing interface over the field catalog
// derived from meta.Registry. Callers do not depend on how definitions are derived.
type FieldCatalog struct {
	byResource map[string]ResourceCatalog
	resources  []string
}

// NewFieldCatalog 从 meta.Registry 派生打印字段目录（单一事实源，对齐旧
// Elixir/Ash 反射口径）：
//   - 目录资源 = 权限前缀的头资源（PrintHead 显式标记；前缀下只有一个非投影
//     候选时自动认定；ReadPermissionsAny 投影视图不参与）；
//   - 头字段 = 自身标量（含计算/投影字段，剔除 id/时间戳/敏感字段/*_id 外键列）
//   - belongs_to 一层展开（relation.目标标量，目标侧跳过计算字段）
//   - 封闭枚举多态的 relation.labelField（如 party.name）；
//   - 循环区 = 头资源 PrintLoops 声明，字段口径与头字段一致，嵌套循环取循环
//     目标自身的 PrintLoops；PrintRawID 多态外键（如子表 party_id）只暴露
//     原始 ID 列；开放字符串多态（voucher）从不展开。
func NewFieldCatalog(registry *meta.Registry) *FieldCatalog {
	if registry == nil {
		panic("打印字段目录需要 meta.Registry")
	}
	catalog := &FieldCatalog{byResource: make(map[string]ResourceCatalog)}
	for _, head := range printHeads(registry) {
		definition := ResourceCatalog{
			Resource: head.PermissionPrefix,
			Fields:   deriveFields(registry, head),
		}
		for _, declared := range head.PrintLoops {
			target, ok := registry.Get(declared.Resource)
			if !ok {
				panic(fmt.Sprintf("打印循环区 %s.%s 指向未知 Meta 资源: %s",
					head.Name, declared.Name, declared.Resource))
			}
			loop := Loop{
				Name:   declared.Name,
				Label:  declared.Name,
				Fields: deriveFields(registry, target),
			}
			for _, nested := range target.PrintLoops {
				loop.NestedLoops = append(loop.NestedLoops, nested.Name)
			}
			sort.Strings(loop.NestedLoops)
			definition.Loops = append(definition.Loops, loop)
		}
		sort.Slice(definition.Loops, func(i, j int) bool {
			return definition.Loops[i].Name < definition.Loops[j].Name
		})
		catalog.byResource[definition.Resource] = definition
		catalog.resources = append(catalog.resources, definition.Resource)
	}
	sort.Strings(catalog.resources)
	return catalog
}

// printHeads 选出每个权限前缀的打印头资源。
func printHeads(registry *meta.Registry) []meta.ResourceMeta {
	candidates := make(map[string][]meta.ResourceMeta)
	marked := make(map[string][]meta.ResourceMeta)
	for _, resource := range registry.Resources() {
		if len(resource.ReadPermissionsAny) > 0 {
			continue
		}
		candidates[resource.PermissionPrefix] = append(candidates[resource.PermissionPrefix], resource)
		if resource.PrintHead {
			marked[resource.PermissionPrefix] = append(marked[resource.PermissionPrefix], resource)
		}
	}
	heads := make([]meta.ResourceMeta, 0, len(candidates))
	for prefix, group := range candidates {
		explicit := marked[prefix]
		switch {
		case len(explicit) > 1:
			panic("权限前缀 " + prefix + " 存在多个打印头资源")
		case len(explicit) == 1:
			heads = append(heads, explicit[0])
		case len(group) == 1:
			heads = append(heads, group[0])
		default:
			names := make([]string, 0, len(group))
			for _, resource := range group {
				names = append(names, resource.Name)
			}
			panic(fmt.Sprintf("权限前缀 %s 的打印头资源不明确（%s），请用 PrintHead 标记",
				prefix, strings.Join(names, ", ")))
		}
	}
	return heads
}

var technicalFields = map[string]struct{}{"id": {}, "inserted_at": {}, "updated_at": {}}

// deriveFields 派生单资源的可打印字段面（头字段与循环区字段同口径）。
func deriveFields(registry *meta.Registry, resource meta.ResourceMeta) []Field {
	names := make(map[string]struct{})
	for _, field := range resource.Fields {
		if _, technical := technicalFields[field.Name]; technical || field.Sensitive {
			continue
		}
		if field.Ref != nil {
			deriveRefFields(registry, resource, field, names)
			continue
		}
		if field.Type == meta.TypeFK || strings.HasSuffix(field.Name, "_id") {
			continue
		}
		names[field.Name] = struct{}{}
	}
	return sortedFields(names)
}

func deriveRefFields(
	registry *meta.Registry,
	resource meta.ResourceMeta,
	field meta.FieldMeta,
	names map[string]struct{},
) {
	ref := field.Ref
	if len(ref.Variants) > 0 {
		// 封闭枚举多态（如 party）：展开 relation.labelField；PrintRawID
		// （如子表 party_id）只暴露原始 ID 列。开放字符串多态（voucher）
		// 字段面不固定，从不展开。
		if ref.DiscriminatorType == nil || *ref.DiscriminatorType != "enum" {
			return
		}
		if field.PrintRawID {
			names[field.Name] = struct{}{}
			return
		}
		for _, variant := range ref.Variants {
			names[refPrefix(field)+"."+variant.LabelField] = struct{}{}
		}
		return
	}
	if ref.Resource == nil {
		return
	}
	target, ok := registry.Get(*ref.Resource)
	if !ok {
		panic(fmt.Sprintf("Meta 资源 %s 字段 %s 的打印关联指向未知资源: %s",
			resource.Name, field.Name, *ref.Resource))
	}
	prefix := refPrefix(field)
	for _, targetField := range target.Fields {
		if _, technical := technicalFields[targetField.Name]; technical ||
			targetField.Sensitive || targetField.Calculated {
			continue
		}
		if targetField.Ref != nil || targetField.Type == meta.TypeFK ||
			strings.HasSuffix(targetField.Name, "_id") {
			continue
		}
		names[prefix+"."+targetField.Name] = struct{}{}
	}
}

// refPrefix 计算关联展开路径前缀：优先 Relation 名（camelCase 转 snake_case），
// 否则取外键列去掉 _id 后缀。
func refPrefix(field meta.FieldMeta) string {
	if field.Ref.Relation != nil && *field.Ref.Relation != "" {
		return snakeCase(*field.Ref.Relation)
	}
	return strings.TrimSuffix(field.DBColumn, "_id")
}

func snakeCase(value string) string {
	var result strings.Builder
	for index, r := range value {
		if r >= 'A' && r <= 'Z' {
			if index > 0 {
				result.WriteByte('_')
			}
			result.WriteRune(r + ('a' - 'A'))
			continue
		}
		result.WriteRune(r)
	}
	return result.String()
}

func sortedFields(names map[string]struct{}) []Field {
	result := make([]Field, 0, len(names))
	for name := range names {
		result = append(result, Field{Name: name, Label: name})
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Name < result[j].Name })
	return result
}

func (c *FieldCatalog) Resources() []string {
	return append([]string(nil), c.resources...)
}

func (c *FieldCatalog) Get(resource string) (ResourceCatalog, bool) {
	definition, ok := c.byResource[resource]
	if !ok {
		return ResourceCatalog{}, false
	}
	return cloneResourceCatalog(definition), true
}

func (c *FieldCatalog) ValidatePlaceholders(resource string, placeholders PlaceholderSet) error {
	definition, ok := c.byResource[resource]
	if !ok {
		return apierror.Validation(
			"不支持的资源类型 "+resource,
			map[string][]string{"resource": {"不在打印字段目录中"}},
		)
	}
	head := fieldSet(definition.Fields)
	loops := make(map[string]Loop, len(definition.Loops))
	for _, loop := range definition.Loops {
		loops[loop.Name] = loop
	}

	unknownHead := make([]string, 0)
	for _, name := range placeholders.Fields {
		if _, exists := head[name]; !exists {
			unknownHead = append(unknownHead, name)
		}
	}
	unknownLoop := make([]string, 0)
	deep := make([]string, 0)
	nestedLoop := make([]string, 0)
	for prefix, suffixes := range placeholders.Nested {
		if loop, exists := loops[prefix]; exists {
			allowed := fieldSet(loop.Fields)
			allowed["_seq"] = struct{}{}
			nestedNames := make(map[string]struct{}, len(loop.NestedLoops))
			for _, name := range loop.NestedLoops {
				nestedNames[name] = struct{}{}
			}
			for _, suffix := range suffixes {
				first := firstSegment(suffix)
				if _, nested := nestedNames[first]; nested {
					nestedLoop = append(nestedLoop, prefix+"."+first)
				} else if _, allowedField := allowed[suffix]; !allowedField {
					unknownLoop = append(unknownLoop, prefix+"."+suffix)
				}
			}
			continue
		}
		for _, suffix := range suffixes {
			full := prefix + "." + suffix
			if strings.Contains(suffix, ".") {
				deep = append(deep, full)
			} else if _, exists := head[full]; !exists {
				unknownHead = append(unknownHead, full)
			}
		}
	}

	parts := make([]string, 0, 4)
	parts = appendErrorPart(parts, "未知头字段", unknownHead)
	parts = appendErrorPart(parts, "未知循环区字段", unknownLoop)
	parts = appendErrorPart(parts, "关联路径只支持一层", deep)
	parts = appendErrorPart(parts, "不支持嵌套循环", nestedLoop)
	if len(parts) == 0 {
		return nil
	}
	message := strings.Join(parts, "；")
	return apierror.Validation(message, map[string][]string{"fileId": {message}})
}

func cloneResourceCatalog(value ResourceCatalog) ResourceCatalog {
	result := value
	result.Fields = append([]Field(nil), value.Fields...)
	result.Loops = make([]Loop, len(value.Loops))
	for index, loop := range value.Loops {
		result.Loops[index] = loop
		result.Loops[index].Fields = append([]Field(nil), loop.Fields...)
		result.Loops[index].NestedLoops = append([]string(nil), loop.NestedLoops...)
	}
	return result
}

func fieldSet(fields []Field) map[string]struct{} {
	result := make(map[string]struct{}, len(fields))
	for _, field := range fields {
		result[field.Name] = struct{}{}
	}
	return result
}

func firstSegment(value string) string {
	if index := strings.IndexByte(value, '.'); index >= 0 {
		return value[:index]
	}
	return value
}

func appendErrorPart(parts []string, label string, names []string) []string {
	if len(names) == 0 {
		return parts
	}
	names = uniqueSorted(names)
	return append(parts, label+": "+strings.Join(names, ", "))
}

func uniqueSorted(values []string) []string {
	set := make(map[string]struct{}, len(values))
	for _, value := range values {
		set[value] = struct{}{}
	}
	result := make([]string, 0, len(set))
	for value := range set {
		result = append(result, value)
	}
	sort.Strings(result)
	return result
}

File: server/src/platform/printing/catalog.ts
/**
 * 打印字段目录：自 meta.Registry 派生。
 * 对齐 server-go platform/printing/catalog.go。
 */
import { ApiError } from '../http/errors.ts'
import type { Registry } from '../meta/registry.ts'
import type { FieldMeta, ResourceMeta } from '../meta/types.ts'
import type { PlaceholderSet, PrintField, PrintLoop, PrintResourceCatalog } from './types.ts'
import { uniqueSorted } from './xlsx.ts'

const TECHNICAL_FIELDS = new Set(['id', 'inserted_at', 'updated_at'])

export function createFieldCatalog(registry: Registry) {
  if (!registry) throw new Error('打印字段目录需要 meta.Registry')

  const byResource = new Map<string, PrintResourceCatalog>()
  /** 打印前缀 → 打印头的 Meta 资源名：路由据此把客户端 prefix 收敛到 sealed registry */
  const metaNameByPrefix = new Map<string, string>()
  const resourceNames: string[] = []

  for (const head of printHeads(registry)) {
    const definition: PrintResourceCatalog = {
      resource: head.permissionPrefix,
      fields: deriveFields(registry, head),
      loops: [],
    }
    for (const declared of head.printLoops ?? []) {
      const target = registry.get(declared.resource)
      if (!target) {
        throw new Error(
          `打印循环区 ${head.name}.${declared.name} 指向未知 Meta 资源: ${declared.resource}`,
        )
      }
      const loop: PrintLoop = {
        name: declared.name,
        label: declared.name,
        fields: deriveFields(registry, target),
        nestedLoops: (target.printLoops ?? []).map((n) => n.name).sort(),
      }
      definition.loops.push(loop)
    }
    definition.loops.sort((a, b) => a.name.localeCompare(b.name))
    byResource.set(definition.resource, definition)
    metaNameByPrefix.set(definition.resource, head.name)
    resourceNames.push(definition.resource)
  }
  resourceNames.sort()

  function resources(): string[] {
    return [...resourceNames]
  }

  /**
   * 打印前缀 → sealed registry 的资源名（未收录返回 undefined）。
   * 打印是「请求形态派生动作码」（spec S9）：路由先经此把客户端 prefix 解析成
   * 目录内资源，再派生动作走 guard——杜绝客户端提供任意 prefix 的路径。
   */
  function resourceNameOf(prefix: string): string | undefined {
    return metaNameByPrefix.get(prefix)
  }

  function get(resource: string): PrintResourceCatalog | undefined {
    const definition = byResource.get(resource)
    return definition ? clonePrintResourceCatalog(definition) : undefined
  }

  function validatePlaceholders(resource: string, placeholders: PlaceholderSet): void {
    const definition = byResource.get(resource)
    if (!definition) {
      throw ApiError.validation(`不支持的资源类型 ${resource}`, {
        resource: ['不在打印字段目录中'],
      })
    }
    const head = fieldSet(definition.fields)
    const loops = new Map(definition.loops.map((loop) => [loop.name, loop]))

    const unknownHead: string[] = []
    for (const name of placeholders.fields) {
      if (!head.has(name)) unknownHead.push(name)
    }
    const unknownLoop: string[] = []
    const deep: string[] = []
    const nestedLoop: string[] = []
    for (const [prefix, suffixes] of Object.entries(placeholders.nested)) {
      const loop = loops.get(prefix)
      if (loop) {
        const allowed = fieldSet(loop.fields)
        allowed.add('_seq')
        const nestedNames = new Set(loop.nestedLoops ?? [])
        for (const suffix of suffixes) {
          const first = firstSegment(suffix)
          if (nestedNames.has(first)) {
            nestedLoop.push(`${prefix}.${first}`)
          } else if (!allowed.has(suffix)) {
            unknownLoop.push(`${prefix}.${suffix}`)
          }
        }
        continue
      }
      for (const suffix of suffixes) {
        const full = `${prefix}.${suffix}`
        if (suffix.includes('.')) {
          deep.push(full)
        } else if (!head.has(full)) {
          unknownHead.push(full)
        }
      }
    }

    const parts: string[] = []
    appendErrorPart(parts, '未知头字段', unknownHead)
    appendErrorPart(parts, '未知循环区字段', unknownLoop)
    appendErrorPart(parts, '关联路径只支持一层', deep)
    appendErrorPart(parts, '不支持嵌套循环', nestedLoop)
    if (parts.length === 0) return
    const message = parts.join('；')
    throw ApiError.validation(message, { fileId: [message] })
  }

  return { resources, resourceNameOf, get, validatePlaceholders }
}

export type FieldCatalog = ReturnType<typeof createFieldCatalog>

function printHeads(registry: Registry): ResourceMeta[] {
  const candidates = new Map<string, ResourceMeta[]>()
  const marked = new Map<string, ResourceMeta[]>()
  for (const resource of registry.list()) {
    if (resource.authz?.readAnyOf && resource.authz.readAnyOf.length > 0) continue
    const group = candidates.get(resource.permissionPrefix) ?? []
    group.push(resource)
    candidates.set(resource.permissionPrefix, group)
    if (resource.printHead) {
      const m = marked.get(resource.permissionPrefix) ?? []
      m.push(resource)
      marked.set(resource.permissionPrefix, m)
    }
  }
  const heads: ResourceMeta[] = []
  for (const [prefix, group] of candidates) {
    const explicit = marked.get(prefix) ?? []
    if (explicit.length > 1) {
      throw new Error(`权限前缀 ${prefix} 存在多个打印头资源`)
    }
    if (explicit.length === 1) {
      heads.push(explicit[0]!)
      continue
    }
    if (group.length === 1) {
      heads.push(group[0]!)
      continue
    }
    const names = group.map((r) => r.name).join(', ')
    throw new Error(
      `权限前缀 ${prefix} 的打印头资源不明确（${names}），请用 PrintHead 标记`,
    )
  }
  return heads
}

function deriveFields(registry: Registry, resource: ResourceMeta): PrintField[] {
  const names = new Set<string>()
  for (const field of resource.fields) {
    if (TECHNICAL_FIELDS.has(field.name) || field.sensitive) continue
    if (field.ref) {
      deriveRefFields(registry, resource, field, names)
      continue
    }
    if (field.type === 'fk' || field.name.endsWith('_id')) continue
    names.add(field.name)
  }
  return sortedFields(names)
}

function deriveRefFields(
  registry: Registry,
  resource: ResourceMeta,
  field: FieldMeta,
  names: Set<string>,
): void {
  const ref = field.ref
  if (!ref) return
  if (ref.variants && ref.variants.length > 0) {
    if (ref.discriminatorType !== 'enum') return
    if (field.printRawId) {
      names.add(field.name)
      return
    }
    for (const variant of ref.variants) {
      names.add(`${refPrefix(field)}.${variant.labelField}`)
    }
    return
  }
  if (!ref.resource) return
  const target = registry.get(ref.resource)
  if (!target) {
    throw new Error(
      `Meta 资源 ${resource.name} 字段 ${field.name} 的打印关联指向未知资源: ${ref.resource}`,
    )
  }
  const prefix = refPrefix(field)
  for (const targetField of target.fields) {
    if (
      TECHNICAL_FIELDS.has(targetField.name) ||
      targetField.sensitive ||
      targetField.calculated
    ) {
      continue
    }
    if (
      targetField.ref ||
      targetField.type === 'fk' ||
      targetField.name.endsWith('_id')
    ) {
      continue
    }
    names.add(`${prefix}.${targetField.name}`)
  }
}

function refPrefix(field: FieldMeta): string {
  if (field.ref?.relation) return snakeCase(field.ref.relation)
  return field.dbColumn.replace(/_id$/, '')
}

function snakeCase(value: string): string {
  let result = ''
  for (let index = 0; index < value.length; index++) {
    const r = value[index]!
    if (r >= 'A' && r <= 'Z') {
      if (index > 0) result += '_'
      result += r.toLowerCase()
    } else {
      result += r
    }
  }
  return result
}

function sortedFields(names: Set<string>): PrintField[] {
  return [...names]
    .sort()
    .map((name) => ({ name, label: name }))
}

function clonePrintResourceCatalog(value: PrintResourceCatalog): PrintResourceCatalog {
  return {
    resource: value.resource,
    fields: value.fields.map((f) => ({ ...f })),
    loops: value.loops.map((loop) => ({
      ...loop,
      fields: loop.fields.map((f) => ({ ...f })),
      nestedLoops: loop.nestedLoops ? [...loop.nestedLoops] : undefined,
    })),
  }
}

function fieldSet(fields: PrintField[]): Set<string> {
  return new Set(fields.map((f) => f.name))
}

function firstSegment(value: string): string {
  const index = value.indexOf('.')
  return index >= 0 ? value.slice(0, index) : value
}

function appendErrorPart(parts: string[], label: string, names: string[]): void {
  if (names.length === 0) return
  parts.push(`${label}: ${uniqueSorted(names).join(', ')}`)
}
